File: day2/puzzle2.go
package main

import (
	"fmt"
	"os"
)

func main() {
	var intCodesFile []int
	pointer := 0
	calcOnePointer := 0
	calcTwoPointer := 0
	resultPointer := 0

	file, err := os.Open("./input.txt")
	if err != nil {
		fmt.Println(err.Error())
	}
	for {
	    if n, _ := fmt.Fscanf(file, "%d,", &pointer); n != 1 {break}
	    intCodesFile = append(intCodesFile,pointer)
	}
	file.Close()
	intCodes := make([]int, len(intCodesFile))

	for i:=0; i<=99; i++ {
	for j:=0; j<=99; j++ {
		copy(intCodes, intCodesFile)
		intCodes[1]=i;
		intCodes[2]=j;
		pointer = 0

		for intCodes[pointer] != 99 {
			if intCodes[pointer] == 1 {
				resultPointer = intCodes[pointer+3]
				if resultPointer <= len(intCodes) {
					calcOnePointer = intCodes[pointer+1]
					calcTwoPointer = intCodes[pointer+2]
					if calcOnePointer <= len(intCodes) && calcTwoPointer <= len(intCodes) {
						intCodes[resultPointer] = intCodes[calcOnePointer] + intCodes[calcTwoPointer]
					}
				}
				pointer += 4
			} else if intCodes[pointer] == 2 {
				resultPointer = intCodes[pointer+3]
				if resultPointer <= len(intCodes) {
					calcOnePointer = intCodes[pointer+1]
					calcTwoPointer = intCodes[pointer+2]
					if calcOnePointer <= len(intCodes) && calcTwoPointer <= len(intCodes) && calcTwoPointer > 0 {
						intCodes[resultPointer] = intCodes[calcOnePointer] * intCodes[calcTwoPointer]
					}
				}
				pointer += 4
			} else {
				pointer++
			}
		}
		if intCodes[0]==19690720 {
			fmt.Println(intCodes)
		}
	}
	}
}

File: day2/puzzle2.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    for i in 0..100{
        for j in 0..100{
            if let Ok(lines) = read_lines("./rust_input.txt") {
                for line in lines {
                    if let Ok(intcode_line) = line {
                        let mut intcodes: Vec<i32> = intcode_line
                            .split(',')
                            .map(|s| s.parse().expect("parse error"))
                            .collect();
                        intcodes[1] = i;
                        intcodes[2] = j;
                        let outcodes = run_intcodes(intcodes);
                        if outcodes[0] == 19690720 {
                            println!("Noun -> {}, Verb -> {}", outcodes[1], outcodes[2]);
                            break;
                        }
                    }
                }
            }
        }
    }
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn run_intcodes(mut ic: std::vec::Vec<i32>) -> std::vec::Vec<i32> {
    let mut pointer = 0;
    loop{
        if ic[pointer] == 99 {
            return ic;
        } else if ic[pointer] == 1 {
            let arg1 = ic[pointer+1] as usize;
            let arg2 = ic[pointer+2] as usize;
            let answer = ic[pointer+3] as usize;
            ic[answer] = ic[arg1] + ic[arg2];
            pointer += 4;
        } else if ic[pointer] == 2 {
            let arg1 = ic[pointer+1] as usize;
            let arg2 = ic[pointer+2] as usize;
            let answer = ic[pointer+3] as usize;
            ic[answer] = ic[arg1] * ic[arg2];
            pointer += 4;
        } else {
            println!("Invalid opcode -> {}", ic[pointer]);
            return ic;
        }
    }
}